Develop into master

rineless opened the pull request.

rineless committed:
Complete initial pack of student data

rineless committed:
Add CSV_Reader with initial structure

rineless committed:
Add classes GroupData and StudentData

rineless committed:
Add classes Group and Student

rineless committed:
Change jdk version from 13 into 11

rineless committed:
Implement method getDataAsList
Implement method getDataFromLine

rineless committed:
Change exception warning in method getLines

rineless committed:
Add initial structure into StudentData class

rineless committed:
Fix typo in path

rineless committed:
Fix #5
Add initial structure into GroupData class

rineless committed:
Rename CSV_Reader into FileReader

rineless committed:
Add abstract class Reader which is extended with FileReader

rineless committed:
Delete GroupData and StudentData classes

rineless committed:
Create interface LineParser

rineless committed:
Create CSVParser which implements LineParser interface

rineless committed:
Fix #5

rineless committed:
Group *Parser and *Reader into util package

rineless committed:
Group Student and Group into model package

rineless committed:
Group StudentData and StudentGroupData into data package

rineless committed:
Add attributes and constructor to Student class

rineless committed:
Add attributes and constructor to Group class

rineless committed:
Modify methods parseLineToStudent and parseLineToGroup in LineParser

rineless committed:
Rename LineParser into ILineParser

rineless committed:
Create SeparatedValuesParser
To parse line with any separator into String array

rineless committed:
Set CSVParser as child from SeparatedValuesParser

rineless committed:
Delete DSVParser
Used just in one place. Can be replaced with SeparatedValuesParser

rineless committed:
Add getter methods for Student class

rineless committed:
Add getter methods for Group class

rineless committed:
Fix #5
Fix #6
Align code

rineless committed:
Create enum Abbreviation for Group

rineless committed:
Create packages group and student

rineless committed:
Modify Abbreviation enum
Create attribut for full name, constructor and getter

rineless committed:
Modify method setOnlineAccess in Group class

rineless committed:
Delete constructor in Group class

rineless committed:
Create enum Gender for student

rineless committed:
Create enum TypeOfStudying for student

rineless committed:
Fix typo in TypeOfStudying

rineless committed:
Create enum TypeOfContract for student

rineless committed:
Create setter for Student

rineless committed:
Delete constructor in Student class

rineless committed:
Fixed #5
Fixed #6
Modify methods parseLineToStudent and parseLineToGroup in ILineParser

rineless committed:
Clean application class

rineless committed:
Unite methods receiveLinesFromFileAsStream and receiveLinesAsList in method receiveLinesAsList.

Method receiveLinesFromFileAsStream was used just in receiveLinesAsList method, so can be neatly deleted

rineless committed:
Modify receiveLinesAsList method

rineless committed:
Delete unnecessary attribut path

rineless committed:
Remove static modifier in receiveLinesAsList methode in FileReader

rineless committed:
Add some new checks of argument into receiveLinesAsList method in FileReader

rineless committed:
Add method to parse String array to line in ILineParser

rineless committed:
Add method to parse Student to line in ILineParser

rineless committed:
Implement method parseArrayToLine of ILineParser in SeparatedValuesParser

rineless committed:
Override toString method for Student

rineless committed:
Fix typo in setGender method in Student

rineless committed:
Rename parameter in constructor for SeparatedValuesParser into separator_regex

rineless committed:
Fix setAbbreviation method in Group by securing uppercase abbreviation

rineless committed:
Add space into argument of parseLineToArray method in setResponsibleForGroup method in Group

rineless committed:
Rename isOnlineAccess method into getOnlineAccess

rineless committed:
Create method parse Group to line in ILineParser

rineless committed:
Override toString method for Group

rineless committed:
Fix mistake in setResponsibleForGroup method in Group

rineless committed:
Simplify "if else" in setOnlineAccess method in Group

rineless committed:
Set right separator_regex for SeparatedValuesParser in setBirthDate method in Student

rineless committed:
Create method receiveListOfStudents in Application

rineless committed:
Create method receiveListOfGroups in Application

rineless committed:
Change all constant names to uppercase in Abbreviation

rineless committed:
Create addStudent method in Application

rineless committed:
Delete online and present attendees columns
Will be calculated later in program from students, so is not needed anymore as attribute of Group

rineless committed:
Delete inappropriate space in StudentData

rineless committed:
Modify parseLineToGroup method in ILineParser

rineless committed:
Add path parameter to receiveLinesAsList method in Reader

rineless committed:
Fix #7

rineless committed:
Create package reader in util. Move FileReader and Reader in reader package

rineless committed:
Create package parser in util. Move CSVParser, ILineParser and SeparatedValuesParser in parser package

rineless committed:
Create tests.util.parser package with SeparatedValuesParserTest

rineless committed:
Add some tests into SeparatedValuesParserTest

rineless committed:
Make some structure changes in Tests in SeparatedValuesParserTest

rineless committed:
Add some tests for parseArrayToLine method in SeparatedValuesParserTest

rineless committed:
Fix parseArrayToLine method in SeparatedValuesParser

TetianaIefimenko commented on src/Application.java:
```diff
@@ -1,6 +1,80 @@
-       System.out.println("Hello world");
+
+        System.out.println("Students:");
+        List<Student> studentList = receiveListOfStudents("src/data/StudentData.csv");
```
try to use relative paths. After compilation you don't have such folder structure.

rineless replied:
`Path current = Path.of(".");

        System.out.println("Students:");
        List<Student> studentList = receiveListOfStudents(current.toString() + "/src/data/StudentData.csv");
        studentList.stream().forEach(student -> System.out.println(student.toString()));

        System.out.println("Groups:");
        List<Group> groupList = receiveListOfGroups(current.toString() + "/src/data/StudentGroupData.csv");
        groupList.stream().forEach(group -> System.out.println(group.toString()));`

TetianaIefimenko commented on src/Application.java:
```diff
@@ -1,6 +1,80 @@
+        groupList.stream().forEach(group -> System.out.println(group.toString()));
+
+        List<String> studentsToAdd = new ArrayList<>();
+        studentsToAdd.add("13,Rubel,Bare,m,15.02.1998,German,Bremen,payable,0,present,rubel.bare@gmail.com");
```
to add a student you need to construct a java object and convert it in a string in your code

rineless replied:
`List<Student> studentsToAdd = new ArrayList<>();
        studentsToAdd.add(new Student().setId(13).setName("Rubel").setSurname("Bare").setGender(Gender.MALE)
                .setBirthDate(LocalDate.of(1998,2,15)).setCitizenship("German").setPlaceOfBirth("Bremen")
                .setTypeOfContract(TypeOfContract.PAYABLE).setGroupId(0).setTypeOfStudying(TypeOfStudying.PRESENT).setContactInformation("rubel.bare@gmail.com"));
        addStudents(studentsToAdd,studentList,current.toString() + "/src/data/StudentData.csv");`

TetianaIefimenko commented on src/Application.java:
```diff
@@ -1,6 +1,80 @@
+        FileReader studentReader = new FileReader();
+        List<String> lineList = studentReader.receiveLinesAsList(path);
+        CSVParser parser = new CSVParser();
+        lineList.remove(0);
```
you can do this operation in the stream using `skip()`

rineless replied:
`FileReader studentReader = new FileReader();

        return studentReader.receiveLinesAsList(path).stream()
                .map(line -> new CSVParser().parseLineToStudent(line)).collect(Collectors.toList());`

TetianaIefimenko commented on src/model/group/Abbreviation.java:
```diff
@@ -0,0 +1,15 @@
+package model.group;
+
+public enum Abbreviation {
```
please, rename it `GroupNames` or give another descriptive name

rineless replied:
`public enum GroupNames`

TetianaIefimenko commented on src/model/group/Group.java:
```diff
@@ -0,0 +1,95 @@
+    private String name;
+    private Abbreviation abbreviation;
+    private Locale language;
+    private boolean onlineAccess;
```
boolean variables usually have names `is**`

rineless replied:
`private boolean isOnlineAccessible;`

TetianaIefimenko commented on src/util/parser/SeparatedValuesParser.java:
```diff
@@ -0,0 +1,23 @@
+    public String[] parseLineToArray(String line) {
+        if (line != null)
+            return line.split(separator);
+        return null;
```
please, don't return null, return empty value or throw an exception

rineless replied:
`public String[] parseLineToArray(String line) {
        if (line != null && !line.contentEquals("")) {
            return line.split(separator_regex);
        }
        return new String[]{};
    }`

TetianaIefimenko commented on src/util/reader/FileReader.java:
```diff
@@ -0,0 +1,24 @@
+    //TODO check exceptions, maybe make a bit different
+    public List<String> receiveLinesAsList(String path) {
+        try {
+            if (Files.exists(Path.of(path)) && Files.isReadable(Path.of(path)))
```
The path may specify a directory. Please, add an appropriate check for this case.

rineless replied:
`public List<String> receiveLinesAsList(String path) {
        try {
            if (Files.exists(Path.of(path)) && Files.isRegularFile(Path.of(path))
                    && Files.isReadable(Path.of(path)))`

rineless committed:
Modify parseLineToArray in SeparatedValuesParser.
 Return empty array instead of null

rineless committed:
Modify parseLineToArray in SeparatedValuesParser.
 Return empty array also if line is empty

rineless committed:
Handle null and empty array case in parseArrayToLine method in SeparatedValuesParser
 Return empty array also if line is empty

rineless committed:
Add some negative tests for lineToArray and arrayToLine methods in SeparatedValuesParserTest

rineless committed:
Simplify some tests by parameterized test in SeparatedValuesParserTest

rineless committed:
Change method name into shouldParseArrayToLine in SeparatedValuesParserTest

rineless committed:
Place tests into nested classes ParseLineToArrayTests and ParseArrayToLineTests in SeparatedValuesParserTest

rineless committed:
Make parseLineToArray_ShouldParseNullToEmptyArray method in SeparatedValuesParserTest parametrized

rineless committed:
Make parseArrayToLine_ShouldParseNullToEmptyArray and parseArrayToLine_ShouldParseEmptyArrayToEmptyLine methods in SeparatedValuesParserTest parametrized

rineless committed:
Create ParseLineToStudentTests nested class in SeparatedValuesParserTest

rineless committed:
Fix warning in shouldParseArrayToLine test in SeparatedValuesParserTest

rineless committed:
Supplement some setter methods with methods with other parameters in Student

rineless committed:
Override equals method for Student

rineless committed:
Create shouldParseLineToStudent parameterized test in SeparatedValuesParserTest

rineless committed:
Add some new arguments to shouldParseLineToStudent test in SeparatedValuesParserTest

rineless committed:
Add null case to parseLineToStudent method in ILineParser

rineless committed:
Add IllegalArgumentException into setGender method in student

rineless committed:
Add IllegalArgumentExceptions to setName and setSurname methods in Student

rineless committed:
Add null case to setGender method in Student

rineless committed:
Add null case to setBirthDate method in Student

rineless committed:
Add IllegalArguments to setCitizenship method in Student

rineless committed:
Add IllegalArguments to setPlaceOfBirth method in Student

rineless committed:
Add null case to setTypeOfContract method in Student

rineless committed:
Add null case to setTypeOfStudying method in Student

rineless committed:
Add IllegalArgumentExceptions to setContractInformation method in Student

rineless committed:
Modify setContactInformation method in Student. Empty contactInformation is now allowed

rineless committed:
Rethrow exception in parseLineToStudent method in ILineParser

rineless committed:
Add negative tests for parseLineToStudent method in SeparatedValuesParserTest

rineless committed:
Add one more setAbbreviation method with Abbreviation parameter in Group

rineless committed:
Add one more setLanguage method with Locale parameter in Group

rineless committed:
Add one more setOnlineAccess method with boolean parameter in Group

rineless committed:
Add one more setResponsibleForGroup method with String[] parameter in Group

rineless committed:
Override equals method for Group

rineless committed:
Fix and add exceptions to setResponsibleForGroup method in Group

rineless committed:
Fix typo in setLanguage in Group

rineless committed:
Modify null array elements into "" in parseArrayToLine method in SeparatedValuesParser

rineless committed:
Add some null cases in parseGroupToLine in ILineParser

rineless committed:
Add IllegalArgumentException if argument is null in parseLineToGroup method in ILineParser

rineless committed:
Add Null and Empty cases to some setters in Group

rineless committed:
Add indication IllegalArgumentException to some methods in Group

rineless committed:
Add indication IllegalArgumentException to some methods in Student

rineless committed:
Fix setName method in Group

rineless committed:
Add null and empty string exception in setLanguage method in Group

rineless committed:
Add some tests for parseLineToGroup method

rineless committed:
Add some tests for parseLineToStudent method

rineless committed:
Add null case to parseStudentToLine method in ILineParser

rineless committed:
Add some null and 0 checks for studentData elements to parseStudentToLine method in SeparatedValuesParser

rineless committed:
Add nested class ParseStudentToLineTests with some tests in SeparatedValuesParserTest

rineless committed:
Add null case to parseGroupToLine method in ILineParser

rineless committed:
Add some new cases for groupData elements in parseGroupToLine method in ILineParser

rineless committed:
Modify setResponsibleForGroup and getResponsibleForGroup methods in Group

rineless committed:
Modify setResponsibleForGroup(string) method in Group

rineless committed:
Add empty string case to setResponsibleForGroup(string) method in Group

rineless committed:
Add more checks to groupData elements in parseGroupToLine method in ILineParser

rineless committed:
Fix bug in setResponsibleForGroup(string) method in Group

rineless committed:
Fixed #8

rineless committed:
Change path to relative in Application

rineless committed:
Rename Abbreviation enum as GroupNames

rineless committed:
Remove setter setFullName and attribute fullName. Remains only groupName (abbreviation). For getting full name from abbreviation can be used getFullName method.

rineless committed:
Remove bug in receiveListOfStudents method (forgot skip first line) in Application

rineless committed:
Fixed #16
Add check if path is pointing on file in receiveLinesAsList method

rineless committed:
Create repository package with GroupsRepository and StudentsRepository classes

rineless committed:
Create PathFinder class with method which return string path by given name in folders from given root. Create package finder with this class

rineless committed:
Add signature throws IllegalArgumentEception for findByName method in PathFinder

rineless committed:
Fix bug in findByName method in PathFinder

rineless committed:
Use PathFinder to find path for StudentData.csv and StudentGroupData.csv in Application

rineless committed:
Fix #16

rineless committed:
Create resources package and move data package into resources

rineless committed:
Add more allowed argument cases for setGender method in Student

rineless committed:
Fix #16
Receive path by getClass().getClassLoader().getResource(name) method in Application
